app/models/event.rb

```diff
@@ -11,11 +11,29 @@
 #  updated_at      :datetime         not null
 #  eventable_id    :bigint
 #  organization_id :bigint
+#  user_id         :bigint
 #
 class Event < ApplicationRecord
   scope :for_organization, ->(organization_id) { where(organization_id: organization_id).order(:event_time) }
 
   serialize :data, EventTypes::StructCoder.new(EventTypes::InventoryPayload)
 
   belongs_to :eventable, polymorphic: true
+  belongs_to :user, optional: true
+
+  before_create do
+    self.user_id = PaperTrail.request&.whodunnit
+  end
+
+  after_create_commit do
```

Comment:
I think it is appropriate for this to immediately identify and record issues when they happen, but flagging they we have to beware of the resulting slowdown. For individual actions I don't think we'd ever notice, but we should be on the lookout for batch operations. Also once there is a diff then we'll always have one (for the org/location) and this'll find things every time unless the problems cancel each other out (a finance friend of mine calls this "The God of Offsets").

Reply:
Yep, we definitely should monitor if there's a huge drop in efficiency. I do want to put in a followup PR that relies more heavily on snapshots (which should be done more often) which should speed this up.

And yeah I know that the diffs will keep piling on. The goal is not for this to be a permanent feature, but for us to monitor over time and verify that the events are the accurate interpretation (and to do fixes until they are). Once we're confident in that, we switch over to using them as the source of truth and the diffs no longer need to happen.